components/config_info/src/lib.rs

```diff
@@ -0,0 +1,330 @@
+/// - max. The maximal valid value of target field. Must be a literal value with the same type or
+///        can be coverted to the type by `TryInfo::try_into`.
+/// - max_desc. Describe the maximal valid value as a string.
+/// - options. A list of value that define the set of all valid values.
+/// - skip. Skip generate config info for this field.
+/// - submodule. This field is a submodule.
+///
+///
+/// # Field Type(#[config_info(type= "..")])
+///
+/// this attribue the value type in the config file, all valid options are:
+/// - Number. Represent a numeric value, auto-assigned for all primitive numeric types. Other custom
+///           defined type should implement `From<T>` in which T is one of the primitive numeric types.
+/// - Array. Represent the array type. auto-assigned for `[T, N]` and `Vec<T>` types.
+/// - Stirng. Represent string type. auto-assigned for `String`, `ReadableSize` and `ReadableDuration`.
+///           The generated code always use `FromStr::from_str` to convert str to target type.
+/// - Boolean. Represent the bool type. auto-assigned for `bool`. Custom defined type should implement
+///            `From<bool>` trait.
+/// - Map. Represent the map type. auto-assigned for `HashMap` and `BTreeMap`.
+///
+/// NOTE: The `type` attribute should be explicitly set if target field type is not build-in supported。
+///
+///
+/// # Field Value Bound (#config_info(min = .., max = ..))
+///
+/// The `min` and `max` attribue define the lower and upper bound of target field. The provide literal
+/// value must be the same type with target field or can be converted to target type with `TryInto::try_into`.
+///
+///
+/// # Field Value Bound Description(#config_info(default_desc="..", min_desc = "...", max_desc = "..."))
+///
+/// All the config types in TiKV has implement the `Default` trait, so the `ConfigInfo` marco by default
+/// generate the `default` value by the `Default` implementation. But some config fields' default value is
+/// based on some env parameters(such as the CPU or Memory quota), so output the default value is some what
+/// misleading. In this can, user can explicitly set the `default_desc` attribue to privide a string to
+/// describe the default value choosen rule.
+///
+/// This same rule also applies for the `min_desc` and `max_desc`.
+///
+///
+/// # Config Value Options(#[config_info(opitons= [ .., .. ])])
+/// The `options` attribue define the set of all valid valid for target field. The value of each element
+/// must be either the same type with target field or can be converted to target type with `TryInto::try_into`.
+///
+///
+/// Example:
+/// Please refer the unit test at the bottom of this file.
+///
+pub trait ConfigInfo {
+    type Encoder: Serialize;
+    /// Get encoder that can be serialize with `serde::Serializer`
+    /// with the disappear of `#[config_info(skip)]` field
+    // NOTE: the `input` parameter is use to extract the value_in_file value,
```

Comment:
Why not always use `self`?

Reply:
the `default` value is fetched from `self` , and the `value-in-file` is fetched from the `` which is the value parsed from the actual config when `-config tikv-server.toml` is provided. The current implement only return the `value-in-file` when its value is different from the default value.

Reply:
I presume the reason to not use `Default::default()` is because some default values are generated in `sanitize`? I actually prefer we print out those as well because sometimes it's hard to figure out what the `sanitize` outputs without looking at the code.